src/ucp/core/ucp_am.c

```diff
@@ -359,6 +359,17 @@ ucp_am_fill_first_footer(ucp_am_first_ftr_t *ftr, ucp_request_t *req)
     ftr->total_size = req->send.length;
 }
 
+static UCS_F_ALWAYS_INLINE void ucp_am_put_user_header(ucp_request_t *req)
+{
+    if (!(req->send.msg_proto.am.flags & UCP_AM_SEND_PRIV_FLAG_FREE_HEADER)) {
+        return;
+    }
```

Comment:
I guess this code could be refactored somehow to be called only once then no need for check at all and no need to reset `UCP_AM_SEND_PRIV_FLAG_FREE_HEADER`

Reply:
See my previous discussion with Mikhail about this flag...
If you still think we don't need to check it let me know

Reply:
I think this check should be outside of this function, it does not matter how it's implemented

Reply:
But then I will need to check the same condition in more than one place (each time I call to ucp_am_put_user_header)

Reply:
otherwise the function does some hidden logic, probably I was missleaded by inconcistent naming of functions, internal and external flags (copy, free, user)